File: cmd/listen.go
package cmd

import (
	"context"
	"fmt"
	"net/url"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"
	"github.com/svix/svix-cli/config"
	"github.com/svix/svix-cli/pretty"
	"github.com/svix/svix-cli/relay"
)

type listenCmd struct {
	cmd *cobra.Command
}

func newListenCmd() *listenCmd {
	noLoggingFlagName := "no-logging"
	lc := &listenCmd{}
	lc.cmd = &cobra.Command{
		Use:   `listen localURL (ex. http://localhost:8000/webhook/)`,
		Short: "Forward webhook requests a local url",
		Long: `listen creates an on-the-fly publicly accessible URL for use when testing webhooks.

The cli then acts as a proxy forwarding any requests to the given local URL.
This is useful for testing your webhook server locally without having to open a port or
change any nat configuration on your network.

Example:
	svix listen http://localhost:8000/webhook/

The above command will return you a unique URL and forward any POST requests it receives
to http://localhost:8000/webhook/`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			printer := pretty.NewPrinter(getPrinterOptions(cmd))

			urlStr := args[0]
			url, err := url.Parse(urlStr)
			if err != nil {
				return fmt.Errorf("invalid local url %s", urlStr)
			}
			var token string
			if viper.IsSet("relay_token") {
				token = viper.GetString("relay_token")
			} else {
				token, err = relay.GenerateToken()
				printer.CheckErr(err)
				viper.Set("relay_token", token)
				err := config.Write(viper.AllSettings())
				printer.CheckErr(err)
			}
			noLogging, err := cmd.Flags().GetBool(noLoggingFlagName)
			printer.CheckErr(err)

			client := relay.NewClient(token, url, &relay.ClientOptions{
				DisableSecurity: viper.GetBool("relay_disable_security"),
				RelayDebugUrl:   viper.GetString("relay_debug_url"),
				Logging:         !noLogging,
			})
			client.Listen(context.Background())
			return nil
		},
	}
	lc.cmd.Flags().Bool(noLoggingFlagName, false, "Disables History Logging")
	return lc
}

File: svix-cli2/src/cmds/listen.rs
use crate::config::{get_config_file_path, Config};
use anyhow::{Context, Result};
use clap::Args;

#[derive(Args)]
pub struct ListenArgs {
    /// The local URL to forward webhooks to
    url: url::Url,
    /// Disables History Logging
    #[clap(long)]
    no_logging: bool,
}

impl ListenArgs {
    pub async fn exec(self, cfg: &Config) -> Result<()> {
        let token = match cfg.relay_token.as_ref() {
            None => {
                let token = crate::relay::token::generate_token()?;
                let mut updated_cfg = cfg.clone();
                updated_cfg.auth_token = Some(token.clone());

                let cfg_path = get_config_file_path()?;
                if let Err(e) = updated_cfg.save_to_disk(&cfg_path).context(format!(
                    "failed to save relay token to config file at `{}`",
                    cfg_path.as_os_str().to_str().unwrap_or_default()
                )) {
                    eprintln!("{e}");
                }
                token
            }
            Some(token) => token.clone(),
        };
        crate::relay::listen(
            self.url,
            token,
            !self.no_logging,
            cfg.relay_debug_url.as_deref(),
            cfg.relay_disable_security.unwrap_or_default(),
        )
        .await?;
        Ok(())
    }
}
